cw/cw305/capture.py

```diff
@@ -606,6 +606,113 @@ def capture_sha3_fvsr_key(ot):
             raise RuntimeError(f'Bad digest: {got} != {expected}.')
         yield ret
 
+def capture_sha3_fvsr_batch (ot, ktp, capture_cfg, scope_type):
+    """Captures traces using the 'b' (batch encryption) command and WaveRunner.
+
+    Args:
+      capture_cfg: Dictionary with capture configuration settings.
+      ot: Initialized OpenTitan target.
+      ktp: Key and plaintext generator.
+      scope: Scope to use for capture.
+    """
+
+    key_fixed = bytearray([0x81, 0x1E, 0x37, 0x31, 0xB0, 0x12, 0x0A, 0x78,
+                           0x42, 0x78, 0x1E, 0x22, 0xB2, 0x5C, 0xDD, 0xF9])
+    ot.target.simpleserial_write("t", key_fixed)
+    key = key_fixed
+    random.seed(capture_cfg["batch_prng_seed"])
+    ot.target.simpleserial_write("s", capture_cfg["batch_prng_seed"].to_bytes(4, "little"))
+
+    # Create the ChipWhisperer project.
+    project_file = capture_cfg["project_name"]
+    project = cw.create_project(project_file, overwrite=True)
+    # Capture traces.
+    rem_num_traces = capture_cfg["num_traces"]
+    num_segments_storage = 1
+    sample_fixed = False
+    # cw and waverunner scopes are supported fot batch capture.
+    scope = SCOPE_FACTORY[scope_type](ot, capture_cfg)
+    with tqdm(total=rem_num_traces, desc="Capturing", ncols=80, unit=" traces") as pbar:
+        while rem_num_traces > 0:
+            # Determine the number of traces for this batch and arm the oscilloscope.
+            scope.num_segments = min(rem_num_traces, scope.num_segments_max)
+
+            scope.arm()
+            # Start batch encryption.
+            ot.target.simpleserial_write(
+                "b", scope.num_segments_actual.to_bytes(4, "little")
+            )
+            # Transfer traces
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == scope.num_segments
+            plaintexts = []
+            ciphertexts = []
+            keys = []
+
+            for i in range(scope.num_segments_actual):
+
+                if sample_fixed:
+                    key = key_fixed
+                else:
+                    random_key = ktp.next()[1]
+                    key = random_key
+
+                plaintext = ktp.next()[1]
+
+                ciphertext = pyxkcp.kmac128(key, ktp.key_len,
+                                            np.asarray(plaintext), ktp.text_len,
+                                            ot.target.output_len,
+                                            b'\x00', 0)
```

Comment:
Can you please cross check the data types here? For example, inside ` capture_sha3_fvsr_key()` the key argument to kmac128 is a bytearray whereas here the same type as the return type of `ktp.next()` is used. I'm not sure if that's also bytearray (probably not, as you do an `np.asarray()` for `plaintext`) or something else. The reason I am asking is that previously such mismatches led to issues when analyzing the traces.

Reply:
This is still an open question, can you please have a look at this @vrozic ?

Reply:
`np.asarray()` is indeed not needed. It was needed before when I was using a different PRNG. I've removed it in the latest commit.

Reply:
Great!